File: apidiff/render_text.py
"""Terminal report, grouped breaking -> additive -> cosmetic."""

from statuslogger import StatusLogger

from .model import ADDED, ADDITIVE, BREAKING, CHANGED, COSMETIC, REMOVED, SEVERITY_ORDER

# Leading glyph per delta kind.
MARKER = {ADDED: "+", REMOVED: "-", CHANGED: "~"}

SEVERITY_LOGGER = {
    BREAKING: StatusLogger.error,
    ADDITIVE: StatusLogger.success,
    COSMETIC: StatusLogger.debug,
}


def render(result, show_warnings: bool = True) -> None:
    """Print a diff to the terminal."""
    StatusLogger.title(
        f"{result.product_name} ({result.product_code})  {result.version_from} -> {result.version_to}"
    )

    if not result.deltas:
        StatusLogger.success("No public API changes.")
        _render_warnings(result, show_warnings)
        return

    for severity in SEVERITY_ORDER:
        deltas = result.by_severity(severity)
        if not deltas:
            continue
        log = SEVERITY_LOGGER[severity]
        StatusLogger.notice(f"\n{severity}  ({len(deltas)})")

        current_type = None
        for delta in deltas:
            if delta.type_uid != current_type:
                current_type = delta.type_uid
                StatusLogger.info(f"  {current_type or 'namespaces'}")
            log(f"    {MARKER[delta.kind]} {delta.display}")
            if delta.kind == CHANGED:
                StatusLogger.message(f"        was: {delta.before}")
                StatusLogger.message(f"        now: {delta.after}")
            elif delta.before or delta.after:
                StatusLogger.message(f"        {delta.before or delta.after}")
            for reason in delta.reasons:
                StatusLogger.message(f"        . {reason}")

    summary = result.summary()
    StatusLogger.title(
        f"\nSummary: {summary['breaking']} breaking, {summary['additive']} additive, "
        f"{summary['cosmetic']} cosmetic"
    )
    _render_warnings(result, show_warnings)


def _render_warnings(result, show_warnings: bool) -> None:
    if not show_warnings or not result.warnings:
        return
    StatusLogger.warning(f"\n{len(result.warnings)} warning(s):")
    for warning in result.warnings[:20]:
        StatusLogger.warning(f"  {warning}")
    if len(result.warnings) > 20:
        StatusLogger.warning(f"  ... and {len(result.warnings) - 20} more")

File: apidiff/render-text.mjs
/** render-text.mjs — terminal report, grouped breaking -> additive -> cosmetic (port of render_text.py). */

import { StatusLogger } from "../statuslogger.mjs";
import { ADDED, BREAKING, ADDITIVE, CHANGED, COSMETIC, REMOVED, SEVERITY_ORDER, bySeverity, summary } from "./model.mjs";

/** Leading glyph per delta kind. */
const MARKER = { [ADDED]: "+", [REMOVED]: "-", [CHANGED]: "~" };

const SEVERITY_LOGGER = {
  [BREAKING]: StatusLogger.error,
  [ADDITIVE]: StatusLogger.success,
  [COSMETIC]: StatusLogger.debug,
};

function renderWarnings(result, showWarnings) {
  if (!showWarnings || !result.warnings.length) return;
  StatusLogger.warning(`\n${result.warnings.length} warning(s):`);
  for (const warning of result.warnings.slice(0, 20)) StatusLogger.warning(`  ${warning}`);
  if (result.warnings.length > 20) StatusLogger.warning(`  ... and ${result.warnings.length - 20} more`);
}

/** Print a diff to the terminal. */
export function render(result, showWarnings = true) {
  StatusLogger.title(
    `${result.productName} (${result.productCode})  ${result.versionFrom} -> ${result.versionTo}`,
  );

  if (!result.deltas.length) {
    StatusLogger.success("No public API changes.");
    renderWarnings(result, showWarnings);
    return;
  }

  for (const severity of SEVERITY_ORDER) {
    const deltas = bySeverity(result, severity);
    if (!deltas.length) continue;
    const log = SEVERITY_LOGGER[severity];
    StatusLogger.notice(`\n${severity}  (${deltas.length})`);

    let currentType = null;
    for (const delta of deltas) {
      if (delta.typeUid !== currentType) {
        currentType = delta.typeUid;
        StatusLogger.info(`  ${currentType || "namespaces"}`);
      }
      log(`    ${MARKER[delta.kind]} ${delta.display}`);
      if (delta.kind === CHANGED) {
        StatusLogger.message(`        was: ${delta.before}`);
        StatusLogger.message(`        now: ${delta.after}`);
      } else if (delta.before || delta.after) {
        StatusLogger.message(`        ${delta.before || delta.after}`);
      }
      for (const reason of delta.reasons) StatusLogger.message(`        . ${reason}`);
    }
  }

  const counts = summary(result);
  StatusLogger.title(
    `\nSummary: ${counts.breaking} breaking, ${counts.additive} additive, ${counts.cosmetic} cosmetic`,
  );
  renderWarnings(result, showWarnings);
}
